src/core/Akka/IO/TcpConnection.cs

```diff
@@ -435,692 +258,523 @@ protected void CompleteConnect(IActorRef commander, IEnumerable<Inet.SocketOptio
+                var chunk = data.Slice(offset, toSend);
+                segs.AddRange(chunk.Buffers);
+                batchBytes += toSend;
 
-            public readonly ReadResultType Type;
-            public readonly SocketError Error;
+                if (toSend == remaining)
+                {
+                    // Full write completed
+                    _pendingWrites.Dequeue();
+                    _state = _state with { QueuedBytes = _state.QueuedBytes - w.Data.Count };
+                    _partialWriteOffset = null;
+                    if (w.WantsAck) _sendArgs.PendingAcks.Add((snd, w.Ack));
+                    if (_traceLogging)
+                        Log.Debug($"[TcpConnection] TrySend: completed full write, dequeued. Remaining queue: {_pendingWrites.Count}");
+                }
+                else
+                {
+                    // Partial write, update offset and break
+                    _partialWriteOffset = offset + toSend;
+                    if (_traceLogging)
+                        Log.Debug($"[TcpConnection] TrySend: partial write, will resume at offset {_partialWriteOffset}");
+                    break;
+                }
+            }
 
-            public ReadResult(ReadResultType type, SocketError error)
+            if (segs.Count == 0)
             {
-                Type = type;
-                Error = error;
+                if (_traceLogging)
+                    Log.Debug("[TcpConnection] TrySend: no segments to send (only empty writes encountered)");
+                return;
             }
-        }
-
-        /// <summary>
-        /// Used to transport information to the postStop method to notify
-        /// interested party about a connection close.
-        /// </summary>
-        protected sealed class CloseInformation
-        {
-            /// <summary>
-            /// TBD
-            /// </summary>
-            public ISet<IActorRef> NotificationsTo { get; }
-            public Tcp.Event ClosedEvent { get; }
 
-            public CloseInformation(ISet<IActorRef> notificationsTo, Tcp.Event closedEvent)
-            {
-                NotificationsTo = notificationsTo;
-                ClosedEvent = closedEvent;
+            _sendArgs.BufferList = segs;
+            _state = _state with { IsSending = true };
+            if (_traceLogging)
+                Log.Debug($"[TcpConnection] TrySend: sending {segs.Count} segments, total bytes={batchBytes}");
+            if (!Socket.SendAsync(_sendArgs)) Self.Tell(_sendArgs, Self);
+        }
+        
+        /* ====================================================================*/
+        /*  Shutdown decision                                                  */
+        /* ====================================================================*/
+        private void HandleClose(IActorRef closeSender, ConnectionClosed closeEvent)
```

Comment:
Probably the most complex / important piece of code in the machinery here: dealing with shutdowns. I mostly preserved this from the original code in spirit - because it actually deals with almost everything EXCEPT FOR REPORTING SOCKET CLOSURES correctly, which we do address now.